### PR TITLE
Items us18

### LINKED ISSUE #22: User Story 18, Items Index Page Statistics
As any kind of user on the system
When I visit the items index page ("/items")
I see an area with statistics:
- the top 5 most popular items by quantity purchased, plus the quantity bought
- the bottom 5 least popular items, plus the quantity bought

"Popularity" is determined by total quantity of that item ordered

### PR DESCRIPTION
**Pull request by leahriffell**
closes #22 

**Contributors**
- Driver(s): Leah
- Navigator(s): Priya, Luke, Grant
  
**What was changed:**
- Model methods for top 5 most and least popular items by quantity sold 
- New sections in items index that displays these most and least popular items within their own section
  - Adds quantity purchased to these items in popular and unpopular
- Added more seeds data
- Created 1 to many connection between users and merchants to connect. merchant to merchant employees

**QUESTIONS FOR THE GROUP**
 1. Currently, the "other items" section (items not in top or bottom 5), includes items that are also in the top and bottom popular items. The way I read this US, it don't believe they specify if these items need to be removed in the "other" section. Would you agree? If not, we can revisit this as a group because I'm not sure how to accomplish this with AR.  **confirmed w/ Brian that pop/unpop items do _not_ need to be removed from middle section**

2. Most and least popular products only display if the user is not a merchant. Based on the US, do you think that merchants should also see these sections? If so, I can add them in (if we use the existing methods, it will only include *active* items but the "other" section for merchants includes both inactive _and_ active items).  **it should display for all types of users, including merchants**

### CODE REVIEW

File: app/models/item.rb [hunk 1]
```diff
@@ -16,6 +16,14 @@ def self.active_items
     Item.where(active?:true)
   end
 
+  def self.most_popular_items
+    active_items.select("items.*,sum(quantity) as sum_qty").joins(:item_orders).group(:id).order(sum_qty: :desc).limit(5)
```
Comment (priyapower) [inline app/models/item.rb:20]: Amazing use of active record in the model!

File: app/controllers/items_controller.rb [hunk 2]
```diff
@@ -6,6 +6,8 @@ def index
       @items = @merchant.items
     else
       @items = Item.active_items
+      @popular_items = Item.most_popular_items
```
Comment (priyapower) [inline app/controllers/items_controller.rb:9]: I like that the controller is just holding the array of popular items from the model method!

File: app/views/items/index.html.erb [hunk 3]
```diff
@@ -4,6 +4,27 @@
 <% else %>
   <h1>All Items</h1>
 <% end %>
+
+<% if !@merchant %>
```
Comment (priyapower) [inline app/views/items/index.html.erb:8]: Awesome logic in the index html

File: app/views/items/index.html.erb [hunk 4]
```diff
@@ -23,5 +44,24 @@
         <p>Inactive</p>
       <% end %>
     </section>
-    <% end %>
+  <% end %>
 </section>
+
+<% if !@merchant %>
+  <h2>Least popular</h2>
+  <section class="grid-container least-popular">
```
Comment (priyapower) [inline app/views/items/index.html.erb:52]: Thank you for making distinct sections in this view page

File: app/views/layouts/application.html.erb [hunk 5]
```diff
@@ -27,7 +27,7 @@
           <%= link_to "Merchant Dashboard", "/merchant" %>
           <%= link_to "Cart: #{cart.total_items}", "/cart" %>
         <% else %>
-        <%= link_to "Cart: #{cart.total_items}", "/cart" %>
+          <%= link_to "Cart: #{cart.total_items}", "/cart" %>
```
Comment (priyapower) [inline app/views/layouts/application.html.erb:30]: This change is just for indenting

File: db/seeds.rb [hunk 6]
```diff
@@ -6,20 +6,54 @@
 #   movies = Movie.create([{ name: 'Star Wars' }, { name: 'Lord of the Rings' }])
 #   Character.create(name: 'Luke', movie: movies.first)
 
-Merchant.destroy_all
-Item.destroy_all
+ItemOrder.destroy_all
```
Comment (priyapower) [inline db/seeds.rb:9]: We have lots of seeds. Should we spend time today fixing these and testing our rails server? If they look good we can transfer them to the Heroku database for deployment!

File: spec/features/items/index_spec.rb [hunk 7]
```diff
@@ -3,60 +3,122 @@
...
-        expect(page).to have_css("img[src*='#{@pull_toy.image}']")
+      within "#item-#{@item_1.id}" do
+        expect(page).to have_css("img[src*='#{@item_1.image}']")
         click_link('item-image')
       end
-        expect(current_path).to eq("/items/#{@pull_toy.id}")
+        expect(current_path).to eq("/items/#{@item_1.id}")
+    end
+
+    it "can see an area with most popular items and their qty purchased" do
```
Comment (priyapower) [inline spec/features/items/index_spec.rb:91]: Awesome TDD here

File: spec/features/items/index_spec.rb [hunk 8]
```diff
@@ -3,60 +3,122 @@
...
+        expect(page).to have_link(@item_4.name)
+        expect(page).to have_content("Amount Purchased: 40")
+        expect(page).to have_link(@item_5.name)
+        expect(page).to have_content("Amount Purchased: 50")
+        expect(page).to have_link(@item_6.name)
+        expect(page).to have_content("Amount Purchased: 60")
+      end
     end
+
+    # add test that "other section" only includes items not in most or least popular section *IF* we want to interpret the US as this
```
Comment (priyapower) [inline spec/features/items/index_spec.rb:122]: Is there also a way to sad path test this?

File: spec/models/item_spec.rb [hunk 9]
```diff
@@ -52,13 +52,58 @@
...
 
-    it ".active_items" do
-      expect(Item.active_items).to eq([@chain, @pull_toy])
+    describe ".active_items" do
+      it "can return list of only active items" do
+        expect(Item.active_items).to eq([@item_1, @item_2, @item_3, @item_4, @item_5, @item_6, @item_7, @item_8, @item_9, @item_10, @item_12])
+      end
     end
+
+    describe '.most_popular_items' do
```
Comment (priyapower) [inline spec/models/item_spec.rb:97]: Awesome model tests!

**Review by priyapower — COMMENTED** [review]
Looks great! Amazing job on solving this!

File: app/views/layouts/application.html.erb (hunk 5 above)
Comment (leahriffell) [reply app/views/layouts/application.html.erb:30]: Done!

File: app/controllers/items_controller.rb [hunk 10]
```diff
@@ -4,9 +4,12 @@ def index
     if params[:merchant_id]
       @merchant = Merchant.find(params[:merchant_id])
       @items = @merchant.items
-    else
-      @items = Item.active_items
+      @popular_items = @merchant.items.by_popularity(sum_qty: :desc)
```
Comment (GDemps) [inline app/controllers/items_controller.rb:7]: Good refactor to DRY it up!

File: app/models/item.rb [hunk 11]
```diff
@@ -16,6 +16,10 @@ def self.active_items
     Item.where(active?:true)
   end
 
+  def self.by_popularity(*order)
+    active_items.select("items.*,sum(quantity) as sum_qty").joins(:item_orders).group(:id).order(order).limit(5)
```
Comment (GDemps) [inline app/models/item.rb:20]: AR on point

**Review by GDemps — APPROVED** [review]

File: db/seeds.rb (hunk 6 above)
Comment (leahriffell) [reply db/seeds.rb:9]: Yeah good idea! I'll add a ticket to review seeds file and clean up.

File: spec/features/items/index_spec.rb (hunk 8 above)
Comment (leahriffell) [reply spec/features/items/index_spec.rb]: Per our convo with Brian, we don't need to worry about excluding popular and unpopular items from the "other" section

**Pull request merged**